src/platform/cc13x2_26x2/KeyValueStoreManagerImpl.cpp

```diff
@@ -21,14 +21,68 @@
  *          Platform-specific key value storage implementation for CC1352
  */
 
+/* this file behaves like a config.h, comes first */
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
 #include <platform/KeyValueStoreManager.h>
 
+#include <platform/cc13x2_26x2/CC13X2_26X2Config.h>
+
 namespace chip {
 namespace DeviceLayer {
 namespace PersistedStorage {
 
+using namespace ::chip::DeviceLayer::Internal;
+
 KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
 
+/**
+ * Calculate crc16 CCITT from the c string key.
+ *
+ * This could be optimized because key is probably a printable string and not a
+ * binary array. The length is limited to 10 bits because the maximum value of
+ * an NV sub ID is 10 bits long. Consult
+ * `<simplelink_sdk>/source/ti/common/nv/nvocmp.c:NVOCMP_MAXSUBID`.
+ */
+static uint16_t calc_key_10(const char * key)
+{
+    uint16_t ret = 0xFFFF;
+
+    while (*key != '\0')
+    {
+        char x;
+        x = ret >> 0x8 ^ *key++;
+        x ^= x >> 0x4;
+        ret = (ret << 0x8) ^ ((uint16_t)(x << 0xC)) ^ ((uint16_t)(x << 0x5)) ^ ((uint16_t) x);
+    }
+
+    ret &= 0x03FF;
+    return ret;
+}
+
+CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
+                                          size_t offset_bytes) const
+{
+    VerifyOrReturnError(key, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(value, CHIP_ERROR_INVALID_ARGUMENT);
+
+    return CC13X2_26X2Config::ReadKVS(calc_key_10(key), value, value_size, read_bytes_size, offset_bytes);
```

Comment:
How much storage space do we have?
I wonder if we could store both key and value so that on 'Get' (and maybe also put) we can validate for colisions.

For 10 bits I believe we have about 50% of collisions when using around 38 different keys.

Reply:
There should be a 40% chance of collision with 511 keys, (2^9). you're correct, the underlying driver was not really made for this, I can increase the keysize by about 8 bits by using the major key bits.

The key size of the key is not a problem for storage, but concatenating them would have to be done in memory because appending to an NV item is not supported. Alloc'ing a certificate plus the key would not be fun.

Reply:
Being able to have a 16 or 18-bit key seems like that would make this a much more workable solution, at least in the near term.  And the more elaborate solutions (introducing the available filesystem, incurring the resulting increased flash space and refactoring other code to use this, OR reallocing buffers to embed the key) definitely present tough tradeoffs.

Is it difficult to add use of the major key bits?  That seems like a good compromise to get something working immediately.

Reply:
I am actually not sufficiently up to date with birthday collisoion formulas (https://en.wikipedia.org/wiki/Birthday_problem) so I got my stats for 50% collision via a python script.

My point is: I believe there is a high chance of collision, much faster than number of available keys will indicate. I assume keys behave like rand(1, 1024) for our estimation purposes.

Reply:
I wonder if the platform could provide some 'preallocated numbers map' instead, that requires manually adding keys and would fail otherwise. This would ensure no colision.

Reply:
This is my test:

```python
import random
from collections import Counter

def test(l):
   data = random.choices(range(1024), k=l)
   return len(data) == len(set(data))

print(Counter([test(38)  for x in range(10000)]))
print(Counter([test(511)  for x in range(10000)]))
```

511 random keys seem to have near 100% colision chance.

Reply:
To be specific:

```python
x = 1
for i in range(39):
     x = x * (1024-i)/1024.0
print(x) # 0.4805249249401561

x = 1
for i in range(512):
     x = x * (1024-i)/1024.0
print(x) # 8.299705841330013e-69
```

Reply:
> I wonder if the platform could provide some 'preallocated numbers map' instead, that requires manually adding keys and would fail otherwise. This would ensure no colision.

That was the original design before the KeyValueStorage was added. 

I must have been confused on the birthday problem.

Reply:
In this case we would need a map of `string -> integer` if we keep KVS "char *" support.

Not sure if this is practical though: I believe for node data, we format things like "prefix_<nodeid>" as key names currently and given nodeid is random it cannot be preallocated.

Reply:
Hey @andy31415 , I removed the hashing algorithm. It didn't seem like a good solution when everyone pointed out the problems. Now the implementation is a pair of parallel arrays that store the keys and values.